adsrefpipe/tests/unittests/test_db_query.py

```diff
@@ -0,0 +1,388 @@
+                'source_filename': os.path.join(stubdata_dir,'00013.raw'),
+                'resolved_filename': '00013.raw.result',
+                'parser': 'Text',
+            }, {
+                'bibcode': '2020arXiv200400014K',
+                'source_filename': os.path.join(stubdata_dir,'00014.raw'),
+                'resolved_filename': '00014.raw.result',
+                'parser': 'Text',
+            }, {
+                'bibcode': '2020arXiv200400016L',
+                'source_filename': os.path.join(stubdata_dir,'00016.raw'),
+                'resolved_filename': '00016.raw.result',
+                'parser': 'Text',
+            }
+        ]
+
+        # test querying bibcodes
+        bibcodes = ['2020arXiv200400013L', '2020arXiv200400014K', '2020arXiv200400016L']
+        result_got = self.app.query_reference_tbl(bibcode_list=bibcodes)
+        self.assertTrue(result_expected == result_got)
+
+        # test querying filenames
+        filenames = [os.path.join(stubdata_dir,'00013.raw'),
+                     os.path.join(stubdata_dir,'00014.raw'),
+                     os.path.join(stubdata_dir,'00016.raw')]
+        result_got = self.app.query_reference_tbl(source_filename_list=filenames)
+        self.assertTrue(result_expected == result_got)
+
+        # test querying both bibcodes and filenames
+        result_got = self.app.query_reference_tbl(bibcode_list=bibcodes, source_filename_list=filenames)
+        self.assertTrue(result_expected == result_got)
+
+        # test if nothing is passed, which return 10 records max
+        result_got = self.app.query_reference_tbl()
+        self.assertTrue(result_expected == result_got)
+
+    def test_query_reference_tbl_when_non_exits(self):
+        """ verify non existence reference record """
+        self.add_stub_data()
+        stubdata_dir = os.path.join(self.test_dir, 'unittests/stubdata')
+
+        # test when bibcode does not exist
+        self.assertTrue(self.app.query_reference_tbl(bibcode_list=['2020arXiv200400015L']) == None)
+
+        # test when filename does not exist
+        self.assertTrue(self.app.query_reference_tbl(source_filename_list=os.path.join(stubdata_dir,'00015.raw')) == None)
+
+        # test when both bibcode and filename are passed and nothing is returned
+        self.assertTrue(self.app.query_reference_tbl(bibcode_list=['2020arXiv200400015L'],
+                                                     source_filename_list=os.path.join(stubdata_dir,'00013.raw')) == None)
+
+    def test_query_reference_tbl_when_empty(self):
+        """ verify reference table being empty """
+        self.assertTrue(self.app.query_reference_tbl() == None)
+
+    def test_insert_reference_record(self):
+        """ test inserting reference record """
+        self.add_stub_data()
+        stubdata_dir = os.path.join(self.test_dir, 'unittests/stubdata')
+
+        # attempt to insert a record that already exists in db
```

Comment:
just to be clear, there are two bibcodes, right? One is associated with file, the other associated with references. my understanding is that bibcode associated with file is unique, but still I made bibcode/filename together be primary key, anticipating the same manuscript have been processed in multiple formats, either by multiple publisher xml parsers, or one produced xml the other raw. so that one is covered. the other bibcode, associated to a reference can actually appear multiple times, it is associated with history. we actually allow a bibcode with low score, specific bibstem/year to get reprocessed for whatever reason user wants.
what that what you are asking?
you know, you are off today Friday, I am off next week, we should have a chat the week of 13th, going over these, reference our memories, and just make sure we have understood each other on each point.